File: lite/examples/video_classification/android/app/src/androidTest/java/org/tensorflow/lite/examples/videoclassification/VideoClassifierTest.kt
/*
 * Copyright 2022 The TensorFlow Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tensorflow.lite.examples.videoclassification

import android.content.Context
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.greaterThan
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.tensorflow.lite.examples.videoclassification.ml.VideoClassifier
import org.tensorflow.lite.support.label.Category
import java.io.*

@RunWith(AndroidJUnit4::class)
class VideoClassificationTest {

    companion object {
        private const val INPUT_FILE_NAME = "carving_ice.mp4"
    }

    private lateinit var videoClassifier: VideoClassifier
    private lateinit var appContext: Context
    private lateinit var testContext: Context

    @Before
    fun setup() {
        appContext = InstrumentationRegistry.getInstrumentation().targetContext
        testContext = InstrumentationRegistry.getInstrumentation().context
        val options =
            VideoClassifier.VideoClassifierOptions.builder().setMaxResult(5)
                .setNumThreads(1)
                .build()
        videoClassifier = VideoClassifier.createFromFileAndLabelsAndOptions(
            appContext,
            "movinet_a0_stream_int8.tflite",
            "kinetics600_label_map.txt",
            options
        )
    }

    @Test
    fun testVideoClassifierWithExampleVideo() {
        // Load the test video.
        val inputStream = testContext.assets.open(INPUT_FILE_NAME)
        val file = createFileFromInputStream(inputStream)
        val media = MediaMetadataRetriever()
        media.setDataSource(file.path)

        // Run classification on all frames in the video
        var categories: List<Category> = listOf()
        val frameCount = media.extractMetadata(
            MediaMetadataRetriever.METADATA_KEY_VIDEO_FRAME_COUNT)!!.toInt()
        for (i in 0 until frameCount) {
            val inputFrame = media
                .getFrameAtIndex(i)
                ?.copy(Bitmap.Config.ARGB_8888, true)
            inputFrame?.let {
                categories = videoClassifier.classify(it)
            }
        }
        media.release()

        // Assert if the top-1 classification result matches with ground truth.
        assertTrue("Classification result isn't empty.", categories.isNotEmpty())
        assertEquals("Top1 category matches.", "carving ice", categories[0].label)
        assertThat("Score is larger than threshold.", categories[0].score, greaterThan(0.6f))
    }

    private fun createFileFromInputStream(inputStream: InputStream): File {
        val f = File("${appContext.cacheDir}/$INPUT_FILE_NAME")
        val outputStream: OutputStream = FileOutputStream(f)
        val buffer = ByteArray(1024)
        var length: Int
        while (inputStream.read(buffer).also { length = it } > 0) {
            outputStream.write(buffer, 0, length)
        }
        outputStream.close()
        inputStream.close()
        return f
    }
}

File: lite/examples/video_classification/raspberry_pi/video_classifier_test.py
# Copyright 2022 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Unit tests for the VideoClassifier wrapper."""

from typing import List
import unittest

import cv2
import numpy as np
from video_classifier import Category
from video_classifier import VideoClassifier
from video_classifier import VideoClassifierOptions

_MODEL_FILE = 'movinet_a0_int8.tflite'
_LABEL_FILE = 'kinetics600_label_map.txt'
_GROUND_TRUTH_LABEL = 'carving ice'
_GROUND_TRUTH_MIN_SCORE = 0.5
_VIDEO_FILE = 'test_data/carving_ice.mp4'
_ALLOW_LIST = ['carving ice', 'sawing wood']
_DENY_LIST = ['chiseling stone']
_SCORE_THRESHOLD = 0.2
_MAX_RESULTS = 3
_ACCEPTABLE_ERROR_RANGE = 0.01


class VideoClassifierTest(unittest.TestCase):

  def setUp(self):
    """Initialize the shared variables."""
    super().setUp()

    # Load frames from the test video.
    cap = cv2.VideoCapture(_VIDEO_FILE)
    frames = []
    for _ in range(int(cap.get(cv2.CAP_PROP_FRAME_COUNT))):
      _, frame = cap.read()
      frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
      frames.append(frame)
    self._frames = frames

  def _run_classification_with_frames(self, classifier: VideoClassifier,
                                      frames: List[np.ndarray]) -> [Category]:
    """Run video classification with all the given frames and return the final result."""
    categories = None
    for frame in frames:
      categories = classifier.classify(frame)

    return categories

  def test_default_option(self):
    """Check if the default option works correctly."""
    classifier = VideoClassifier(_MODEL_FILE, _LABEL_FILE)

    # Run classification with frames from the test video.
    categories = self._run_classification_with_frames(classifier, self._frames)

    # Check if TOP-1 result match.
    top_1_category = categories[0]
    self.assertEqual(
        _GROUND_TRUTH_LABEL, top_1_category.label,
        'Label {0} does not match with ground truth {1}'.format(
            top_1_category.label, _GROUND_TRUTH_LABEL))
    self.assertLessEqual(
        _GROUND_TRUTH_MIN_SCORE, top_1_category.score,
        'Classification score {0} is smaller than threshold {1}'.format(
            top_1_category.score, _GROUND_TRUTH_MIN_SCORE))

  def test_allow_list(self):
    """Test the label_allow_list option."""
    option = VideoClassifierOptions(label_allow_list=_ALLOW_LIST)
    classifier = VideoClassifier(_MODEL_FILE, _LABEL_FILE, option)

    # Run classification with frames from the test video.
    categories = self._run_classification_with_frames(classifier, self._frames)

    for category in categories:
      label = category.label
      self.assertIn(
          label, _ALLOW_LIST,
          'Label "{0}" found but not in label allow list'.format(label))

  def test_deny_list(self):
    """Test the label_deny_list option."""
    option = VideoClassifierOptions(label_deny_list=_DENY_LIST)
    classifier = VideoClassifier(_MODEL_FILE, _LABEL_FILE, options=option)

    # Run classification with frames from the test video.
    categories = self._run_classification_with_frames(classifier, self._frames)

    for category in categories:
      label = category.label
      self.assertNotIn(label, _DENY_LIST,
                       'Label "{0}" found but in deny list.'.format(label))

  def test_score_threshold_option(self):
    """Test the score_threshold option."""
    option = VideoClassifierOptions(score_threshold=_SCORE_THRESHOLD)
    classifier = VideoClassifier(_MODEL_FILE, _LABEL_FILE, options=option)

    # Run classification with frames from the test video.
    categories = self._run_classification_with_frames(classifier, self._frames)

    for category in categories:
      score = category.score
      self.assertGreaterEqual(
          score, _SCORE_THRESHOLD,
          'Classification with score lower than threshold found. {0}'.format(
              category))

  def test_max_results_option(self):
    """Test the max_results option."""
    option = VideoClassifierOptions(max_results=_MAX_RESULTS)
    classifier = VideoClassifier(_MODEL_FILE, _LABEL_FILE, options=option)

    # Run classification with frames from the test video.
    categories = self._run_classification_with_frames(classifier, self._frames)

    self.assertLessEqual(
        len(categories), _MAX_RESULTS, 'Too many results returned.')


if __name__ == '__main__':
  unittest.main()
